packages/tpetra/core/example/Finite-Element-Assembly/fem_assembly_type3_staticprofile.cpp

```diff
@@ -0,0 +1,322 @@
+  // in the same manner as we filled in the graph but in this case, nodes 
+  // associated with each element will receive contributions according to 
+  // the row in this stencil.
+  //
+  // In this example, the calls to sumIntoGlobalValues() on 1 core will look like:
+  //   Element 0
+  // - sumIntoGlobalValues( 0,  [  0  1  5  4  ],  [  2  -1  0  -1  ])
+  // - sumIntoGlobalValues( 1,  [  0  1  5  4  ],  [  -1  2  -1  0  ])
+  // - sumIntoGlobalValues( 5,  [  0  1  5  4  ],  [  0  -1  2  -1  ])
+  // - sumIntoGlobalValues( 4,  [  0  1  5  4  ],  [  -1  0  -1  2  ])
+  // Element 1
+  // - sumIntoGlobalValues( 1,  [  1  2  6  5  ],  [  2  -1  0  -1  ])
+  // - sumIntoGlobalValues( 2,  [  1  2  6  5  ],  [  -1  2  -1  0  ])
+  // - sumIntoGlobalValues( 6,  [  1  2  6  5  ],  [  0  -1  2  -1  ])
+  // - sumIntoGlobalValues( 5,  [  1  2  6  5  ],  [  -1  0  -1  2  ])
+  // Element 2
+  // - sumIntoGlobalValues( 2,  [  2  3  7  6  ],  [  2  -1  0  -1  ])
+  // - sumIntoGlobalValues( 3,  [  2  3  7  6  ],  [  -1  2  -1  0  ])
+  // - sumIntoGlobalValues( 7,  [  2  3  7  6  ],  [  0  -1  2  -1  ])
+  // - sumIntoGlobalValues( 6,  [  2  3  7  6  ],  [  -1  0  -1  2  ])
+  //
+  // Similarly to the Graph construction above, we loop over both local and global
+  // elements and insert rows for only the locally owned rows.
+  //
+  RCP<TimeMonitor> timerElementLoopMatrix = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("3) ElementLoop  (Matrix)")));
+
+  RCP<MatrixType> crs_matrix = rcp(new MatrixType(crs_graph));
+
+  scalar_2d_array_type element_matrix;
+  Kokkos::resize(element_matrix, 4, 4);
+
+  Teuchos::Array<GlobalOrdinal> column_global_ids(4);     // global column ids list
+  Teuchos::Array<Scalar> column_scalar_values(4);         // scalar values for each column
+
+  // Loop over owned elements:
+  for(size_t element_gidx=0; element_gidx<mesh.getNumOwnedElements(); element_gidx++)
+  {
+    // Get the stiffness matrix for this element
+    ReferenceQuad4(element_matrix);
+
+    // Fill the global column ids array for this element
+    for(size_t element_node_idx=0; element_node_idx<owned_element_to_node_ids.extent(1); element_node_idx++)
+    {
+      column_global_ids[element_node_idx] = owned_element_to_node_ids(element_gidx, element_node_idx);
+    }
+    
+    // For each node (row) on the current element:
+    // - populate the values array
+    // - add values to crs_matrix if the row is owned.
+    //   Note: hardcoded 4 here because we're using quads.
+    for(size_t element_node_idx=0; element_node_idx<4; element_node_idx++)
+    { 
+      GlobalOrdinal global_row_id = owned_element_to_node_ids(element_gidx, element_node_idx);
+      if(mesh.nodeIsOwned(global_row_id)) 
+      {
+        for(size_t col_idx=0; col_idx<4; col_idx++)
+        {
+          column_scalar_values[col_idx] = element_matrix(element_node_idx, col_idx);
+        }
+        crs_matrix->sumIntoGlobalValues(global_row_id, column_global_ids, column_scalar_values);
```

Comment:
When we use sumIntoGlobalValues or any equivalent APIs, do we assume that a user intend not to use on-node parallel assembly ?

Reply:
@kyungjoo-kim Not necessarily.  If users promise never to call `sumIntoGlobalValues` for off-process rows, then they could use `sumIntoGlobalValues` to implement the local or total element loop approaches (what we've been calling Type 2 and Type 3).  Some users prefer global indices.